### PR TITLE
mullvad-browser: add macos support

### PR DESCRIPTION
**Pull request by heywoodlh**
## Description of changes

Added MacOS support to mullvad-browser package. I'm happy to help support the MacOS version of this Nix package (will likely try to port the Firefox package to MacOS as well). 😄 

I don't feel like this is significant enough to add release notes, but please let me know if I should.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc: @felschr @panicgh

### CODE REVIEW

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix [hunk 1]
```diff
@@ -52,13 +52,14 @@
 
 , libvaSupport ? mediaSupport
 , libva
+, p7zip
```
Comment (felschr) [inline pkgs/applications/networking/browsers/mullvad-browser/default.nix:55]:
```suggestion

, p7zip
```

**Comment by felschr** [comment]
Result of `nixpkgs-review pr 283874` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mullvad-browser</li>
  </ul>
</details>

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix [hunk 2]
```diff
@@ -88,10 +89,20 @@ let
       ++ lib.optionals pulseaudioSupport [ libpulseaudio ]
       ++ lib.optionals libvaSupport [ libva ]
       ++ lib.optionals mediaSupport [ ffmpeg ]
-  );
+  ) else "";
 
   version = "13.0.9";
 
+  # MacOS sources
+  darwinSource = "https://cdn.mullvad.net/browser/${version}/mullvad-browser-macos-${version}.dmg";
```
Comment (felschr) [inline pkgs/applications/networking/browsers/mullvad-browser/default.nix:97]: What about the other mirrors that were used for the linux package?
I think they should also all distribute the macOS package.

Perhaps we should streamline the source urls this into a function for all platforms.

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix [hunk 3]
```diff
@@ -104,6 +115,8 @@ let
       ];
       hash = "sha256-TAtBlSkfpqsROq3bV9kwDYIJQAXSVkwxQwj3wIYEI7k=";
     };
+    aarch64-darwin = darwinFetch;
+    x86_64-darwin = darwinFetch;
```
Comment (felschr) [inline pkgs/applications/networking/browsers/mullvad-browser/default.nix:119]: Not sure if this works with `update-source-version` in the update script, but I'd hope it is.

**Review by felschr — CHANGES_REQUESTED** [review]
The update script needs some adjustments to support darwin platforms:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/browsers/tor-browser/update.nix

Also, would you be interested in adding darwin support for `tor-browser` as well?
These days the package is very similar (it used to be quite different).

**Comment by felschr** [comment]
@panicgh I'm still getting segfaults when trying to run this on my machine. Can you try if the package still works for you on Linux?

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix [hunk 4]
```diff
@@ -146,7 +170,12 @@ stdenv.mkDerivation rec {
     categories = [ "Network" "WebBrowser" "Security" ];
   })];
 
-  buildPhase = ''
+  buildPhase = if stdenv.isDarwin then ''
```
Comment (felschr) [inline pkgs/applications/networking/browsers/mullvad-browser/default.nix:173]: The `buildPhase` does quite a lot of things for Linux and is basically skipped completely here for macOS.

`touch "$MB_IN_STORE/system-install"`
Tor Browser, at least on Linux, is a portable package and stores its state within its own folder.
Creating this `system-install` file causes it to store its state in a user directory (`~/.mullvad/mullvadbrowser` on Linux) instead. Not sure how this works on macOS, perhaps you could tell me where it stores its state with the current setup.

`FONTCONFIG_FILE=$MB_IN_STORE/fontconfig/fonts.conf`
We need it in the nix package, because we don't invoke the `start-mullvad-browser` wrapper directly.
`start-mullvad-browser` sets up the environment & adds some functionality like `--register-app`/`--unregister-app` and then launches the actual binary. This doesn't fully work on NixOS, so we have our own wrapping where we need to make sure to set important variables like `FONTCONFIG_FILE` to avoid leaking fonts & thus changing the browesr fingerprint.

Then there are a couple of settings, some of those are to avoid capturing nix store paths, some change some defaults like `noscript.firstRunRedirection`.

I hope this already gives you a bit of context about this.
Generally, I think we should try to make the Linux & macOS configurations the same except for OS-specific configuration (pulseaudio, etc.).

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix [hunk 5]
```diff
@@ -235,7 +264,16 @@ stdenv.mkDerivation rec {
...
   '';
 
-  installPhase = ''
+  installPhase = if stdenv.isDarwin then ''
+    runHook preInstall
+
+    mkdir -p "$out/Applications/"
+    cp -r "Mullvad Browser/Mullvad Browser.app" "$out/Applications"
+    chmod +x "$out/Applications/Mullvad Browser.app/Contents/MacOS/mullvadbrowser"
+    makeWrapper "$out/Applications/Mullvad Browser.app/Contents/MacOS/mullvadbrowser" "$out/bin/mullvad-browser"
```
Comment (felschr) [inline pkgs/applications/networking/browsers/mullvad-browser/default.nix:273]: On macOS, we should also ship `distribution/distribution.ini` & `distribution/policies.json`.

Comment (felschr) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix:273]: You can verify that it's working in the Help -> About Mullvad Browser dialog where it'll mention NixOS.
I guess the "NixOS" mentions should be renamed to "Nixpkgs" or similar.

image

**Review by felschr — CHANGES_REQUESTED** [review]

**Comment by heywoodlh** [comment]
@felschr responding here instead of in each conversation for now.

Will be updating my commit with your suggestions. 👍🏻 

> Also, would you be interested in adding darwin support for tor-browser as well?
> These days the package is very similar (it used to be quite different).

Once I get this PR in a suitable state, I'd be happy to do the same thing for Tor Browser! But will only do that _after_ I can mentally be done with this PR. :smile:

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 5 above)
Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix:273]: @felschr so have been digging into this and it seems Firefox has some helpful documentation:

https://support.mozilla.org/en-US/kb/deploying-firefox-customizations-macos

I added this to the installPhase, but was unable to get things working (no matter what, MacOS complained the package was broken):

```
    /usr/bin/xattr -r -d com.apple.quarantine "$out/Applications/Mullvad Browser.app"
    install -Dvm644 ${distributionIni} "$out/Applications/Mullvad Browser.app/Resources/distribution/distribution.ini"
    install -Dvm644 ${policiesJson} "$out/Applications/Mullvad Browser.app/Resources/distribution/policies.json"
```

Tbh, I don't see it being _hugely_ important to provide `distribution.ini` or `policies.json` while having to circumvent Apple's quarantining.

However, if you (or anyone) knows how to get around this, I'd be happy to try other solutions.

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 2 above)
Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: Does this solution seem sensible? https://github.com/NixOS/nixpkgs/pull/283874/commits/fa7ff3c54ae65c205b4e03ee43fe0890640f2005#diff-1bd78721e31ae529d71193e63f92f114354272ba99d60406026dbfb41799b27aR97-R109

Not sure how this will impact/cooperate with the tor browser update.nix

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 3 above)
Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: See https://github.com/NixOS/nixpkgs/pull/283874#discussion_r1468226947

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 2 above)
Comment (felschr) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: I'm not entirely sure about the use of `stdenv` here as that would probably mean `sources` can't be evaluated on any platform anymore. Just to go sure, instead of relying on `stdenv` perhaps just turn it into a function:

```nix
mkUrls = platform: extension: [
  "https://cdn.mullvad.net/browser/${version}/mullvad-browser-${platform}-${version}.${extension}"
  "https://github.com/mullvad/mullvad-browser/releases/download/${version}/mullvad-browser-${platform}-${version}.${extension}"
  "https://archive.torproject.org/tor-package-archive/mullvadbrowser/${version}/mullvad-browser-${platform}-${version}.${extension}"
  "https://dist.torproject.org/mullvadbrowser/${version}/mullvad-browser-${platform}-${version}.${extension}"
  "https://tor.eff.org/dist/mullvadbrowser/${version}/mullvad-browser-${platform}-${version}.${extension}"
  "https://tor.calyxinstitute.org/dist/mullvadbrowser/${version}/mullvad-browser-${platform}-${version}.${extension}"
];
```

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 3 above)
Comment (felschr) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: Once the update script is adjusted we can test if it works.

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 5 above)
Comment (felschr) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: The correct seems to be `/Applications/Mullvad Browser.app/Contents/Resources/distribution/distribution.ini`:
https://github.com/mozilla/dinobuildr/blob/ceb419c13daffa97cd6d7a234af23d9aaf7286a7/resources/firefox-bookmarks.sh#L19

Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: Right, in this case I'm attempting to modify the .app in $out instead of moving Mullvad Browser.app to the privileged, system-wide /Applications folder (which would require elevated privileges to edit and shouldn't be necessary just to execute the binary in the .app).

So for now, would it be reasonable to not include the distribution.ini and policies.json and just leave that section commented out?

I reached out on Matrix to see if anyone has other ideas/suggestions.

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 2 above)
Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: Thank you for the example! Will implement tomorrow 👍

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 5 above)
Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: Looks like someone came up with a potential solution that I will try tomorrow!

https://matrix.to/#/!lheuhImcToQZYTQTuI%3Anixos.org/%246Q-M8yqPp44v3zMYWmnlbJa89mbWivA4Lr58552Luf4


> If you’re repackaging an upstream app bundle, and it’s notarized, making any changes to it will flag it as broken on macOS 13+. (edited)

Comment (felschr) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: > Right, in this case I'm attempting to modify the .app in $out instead of moving Mullvad Browser.app to the privileged, system-wide /Applications folder

What I actually meant is the missing `/Contents/` part in the paths in your code snippet.

Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: Oh, shoot! I didn't even notice, thank you.

Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: It seems that `distribution.ini` has no impact on the MacOS version of the app. I ended up testing on my Homebrew installed Mullvad Browser:

image

image

(However, from that screenshot you can observe that `policies.json` is working)

Comment (felschr) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: Great to hear the `policies.json` is working. The `distribution.ini` we can live without, I think.

Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: I can't figure out how to get the Nix-built .app to cooperate with MacOS' quarantining after adding `distribution.ini` and `policies.json`. I think we should give up on this piece for now as `distribution.ini` seems to not have an impact and I don't think that updating the Nix-built `.app` will do anything since it will live in the Nix store.

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 4 above)
Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix:177]: > touch "$MB_IN_STORE/system-install"
> Tor Browser, at least on Linux, is a portable package and stores its state within its own folder.
Creating this system-install file causes it to store its state in a user directory (~/.mullvad/mullvadbrowser on Linux) instead. Not sure how this works on macOS, perhaps you could tell me where it stores its state with the current setup.

The comparable `~/.mullvad/mullvadbrowser` directory is `~/Library/Application Support/MullvadBrowser` on MacOS:

image

Is this `system-install` behavior documented anywhere? My first assumption is that it doesn't operate the same on MacOS.

---

The other changes you point out in `buildPhase` -- in my mind -- wouldn't translate to MacOS (i.e. since MacOS apps are self-contained, MacOS doesn't handle fonts the same way Linux does, etc.), although, I appreciate the context and its helpful nonetheless. Would you like me to more fully investigate these and try to copy the same changes in the build?

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 5 above)
Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: Is there any alternative location that Firefox/Tor Browser/Mullvad Browser looks for `policies.json` (i.e. `~/.mullvad/mullvadbrowser/policies.json`)? I can't find any documentation on it, but if there was, we could probably make a MacOS wrapper that does something like this:

```
cp ${policiesJson} "~/Library/Application Support/MullvadBrowser/policies.json"
$out/bin/mullvad-browser $@
```

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 4 above)
Comment (felschr) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix:177]: > Is this system-install behavior documented anywhere? My first assumption is that it doesn't operate the same on MacOS.

Not really, I had quite a few issues even finding & understanding that functionality.
I originally found it used by the Flatpak: https://github.com/flathub/net.mullvad.MullvadBrowser/blob/27161259ff26a28efc839c726ebde9a4d4b4ac29/net.mullvad.MullvadBrowser.yml#L67

However you can find some mentions of `system-install` in Tor Project's GitLab including the source code using it:
https://gitlab.torproject.org/search?scope=issues&search=system-install
https://gitlab.torproject.org/search?search=%22system-install%22&nav_source=navbar&project_id=472&group_id=477&search_code=true&repository_ref=tor-browser-115.7.0esr-13.5-1

And these config files for Mullvad Browser & Tor Browser could also be of interest, as they define the directories:
https://gitlab.torproject.org/tpo/applications/mullvad-browser/-/blob/mullvad-browser-115.6.0esr-13.5-1/browser/config/mozconfigs/mullvad-browser?ref_type=heads
https://gitlab.torproject.org/tpo/applications/tor-browser/-/blob/tor-browser-115.7.0esr-13.5-1/browser/config/mozconfigs/tor-browser?ref_type=heads

> The comparable `~/.mullvad/mullvadbrowser` directory is `~/Library/Application Support/MullvadBrowser` on MacOS

Does it already use that directory without the `system-install` file? In that case the `system-install` is probably not necessary.

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 2 above)
Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: Does this work?
https://github.com/NixOS/nixpkgs/blob/46c2609eba835c898d23e3f06097e4b1cdf7c941/pkgs/applications/networking/browsers/mullvad-browser/default.nix#L97-L119


```
  mkUrls = platform: extension: [
    "https://cdn.mullvad.net/browser/${version}/mullvad-browser-${platform}-${version}.${extension}"
    "https://github.com/mullvad/mullvad-browser/releases/download/${version}/mullvad-browser-${platform}-${version}.${extension}"
    "https://archive.torproject.org/tor-package-archive/mullvadbrowser/${version}/mullvad-browser-${platform}-${version}.${extension}"
    "https://dist.torproject.org/mullvadbrowser/${version}/mullvad-browser-${platform}-${version}.${extension}"
    "https://tor.eff.org/dist/mullvadbrowser/${version}/mullvad-browser-${platform}-${version}.${extension}"
    "https://tor.calyxinstitute.org/dist/mullvadbrowser/${version}/mullvad-browser-${platform}-${version}.${extension}"
  ];

  sources = {
    x86_64-linux = fetchurl {
      urls = mkUrls "linux-x86_64" "tar.xz";
      hash = "sha256-TAtBlSkfpqsROq3bV9kwDYIJQAXSVkwxQwj3wIYEI7k=";
    };
    aarch64-darwin = fetchurl {
      urls = mkUrls "macos" "dmg";
      hash = "sha256-1rXzXLbmBd567MCUNGL1Vw0xfpk4WdoAG6weJgz9+Pg=";
    };
    x86_64-darwin = fetchurl {
      urls = mkUrls "macos" "dmg";
      hash = "sha256-1rXzXLbmBd567MCUNGL1Vw0xfpk4WdoAG6weJgz9+Pg=";
    };
  };
```

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 4 above)
Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix:172]: > Does it already use that directory without the system-install file? 

I believe so, but is there any way I can verify this conclusively? I would imagine that the Tor Browser/Mullvad Browser is not as ephemeral/self-contained on MacOS as it is on Linux and as a result uses that directory by default.

Comment (felschr) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix:172]: > The other changes you point out in `buildPhase` -- in my mind -- wouldn't translate to MacOS (i.e. since MacOS apps are self-contained. [...] Would you like me to more fully investigate these and try to copy the same changes in the build?

If possible, it would be nice if `extraPrefs` also works on macOS. And then we could also set the `extensions.xpiState` & `noscript.firstRunRedirection` config values.

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 5 above)
Comment (felschr) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: I don't know if there is another way for this.

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 2 above)
Comment (felschr) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: That's exactly what I imagined.

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 4 above)
Comment (felschr) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix:172]: > I believe so, but is there any way I can verify this conclusively?

If you can see the `profiles.ini` & profile directory in there I think that's proof enough.
I don't remember properly, but on Linux without `system-install` file it would try to store its state in the directory of the binary or the working directory or something like that.
As long as it always stores its state in the same place, I don't see an issue here.

**Comment by felschr** [comment]
It would be nice if we could get another macOS user involved as a reviewer here, since my lack of expertise in that area.

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix (hunk 5 above)
Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: So, to test, I reverted Mullvad Browser to 13.0.7 to see what happened with the update mechanism (and maybe figure out if it would work reasonably without `policies.json`). It attempted to automatically download and install the update and then just didn't do anything. Then after closing the app, it wouldn't re-launch. 

So, getting the .app to work with `policies.json` is an absolute priority before this PR can be merged because the auto-update mechanism breaks Mullvad Browser when built with Nix. :thumbsdown:

Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: Made progress! Seems that code signing is the issue. I am able to launch the modified `Mullvad Browser.app` after removing the code signature with this command:

```
/usr/bin/codesign -fs - "$out/Applications/Mullvad Browser.app"
```

However, the browser isn't functional at all. It launches, but the tabs are all crashed and I can't browse to any websites. I'm sure there's some way to get around this -- but, I'm very happy to find that the issue isn't the file attributes in the app. 😄 

Hopefully a more experienced Nix+MacOS user appears to help me figure out what else I am missing.

Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: I think that at this point, Mullvad Browser's current packaging for MacOS makes it essentially incompatible with Nix. I have filed an upstream issue with Mullvad Browser to (hopefully) address this: https://github.com/mullvad/mullvad-browser/issues/219#issue-2103969667

I will plan on waiting for a response there before continuing with this. I _really_ hope they give my feedback some serious consideration. 🤞🏻

Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: Found this just now: https://support.mozilla.org/en-US/kb/managing-policies-macos-desktops

Gonna give it a try 🤞🏻

Comment (heywoodlh) [reply pkgs/applications/networking/browsers/mullvad-browser/default.nix]: @felschr I discovered that on MacOS you can break autoupdating for Mullvad Browser by making the directory it uses for updates user immutable. The directory in question is `~/Library/Application\ Support/MullvadBrowser/UpdateInfo/`.

Potentially, I could write a wrapper script that does something like this:

```
echo 'Disable auto-updates'
chflags -R uchg "$HOME/Library/Application Support/MullvadBrowser/UpdateInfo/"
"$out/Applications/Mullvad Browser.app/Contents/MacOS/mullvadbrowser"

echo 'Disabling auto-updates'
chflags -R nouchg "$HOME/Library/Application Support/MullvadBrowser/UpdateInfo/"
```

When this directory is immutable, Mullvad Browser fails with an "unknown error" when checking for updates (but the functionality of the app is unaffected).

Would this be an unreasonable workaround to implement? It's kind of hack-ey so my feelings won't be hurt if it's a no-go. :smile:

**Comment by heywoodlh** [comment]
TL;DR I'd like to back out my custom branding and auto-update changes and just let the app work the way Mullvad Browser intends on MacOS

---

@felschr I've been using some other MacOS applications that have been packaged in nixpkgs -- for example, Spotify. These apps do not have any custom branding and don't disable the developer's auto-update mechanism (which, in most cases, ends up placing apps in the `/Applications` folder).

I propose I back out my changes to try to add custom branding and disable auto-updates and just let Mullvad Browser's defaults work. I think this would be worth it in order to package mullvad-browser in nixpkgs -- even though, auto-updates would result in the app not exclusively being managed by nixpkgs.

Would love to get thoughts on this?

**Comment by superherointj** [comment]
Please fix merge conflict.

Checking EditorConfig / tests is also failing.

**Comment by heywoodlh** [comment]
I've backed out the custom changes but not sure why the Application isn't running properly. I'm considering myself blocked, maybe someone with more MacOS knowledge than me can help get me across the finish line.

**Comment by superherointj** [comment]
Still, fix the merge conflict.

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix [hunk 6]
```diff
@@ -259,7 +297,7 @@ stdenv.mkDerivation rec {
     mainProgram = "mullvad-browser";
     homepage = "https://mullvad.net/en/browser";
     platforms = attrNames sources;
-    maintainers = with maintainers; [ felschr panicgh ];
+    maintainers = with maintainers; [ felschr panicgh heywoodlh ];
```
Comment (superherointj) [inline pkgs/applications/networking/browsers/mullvad-browser/default.nix:300]:
```suggestion
    maintainers = with maintainers; [ felschr heywoodlh panicgh ];
```
Nit: alphabetical order.

**Comment by superherointj** [comment]
@heywoodlh ping

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix [hunk 7]
```diff
@@ -53,12 +53,15 @@
...
+, p7zip
+, darwin
+
 # Extra preferences
 , extraPrefs ? ""
 }:
 
 let
-  libPath = lib.makeLibraryPath (
+  libPath = if stdenv.isLinux then lib.makeLibraryPath (
```
Comment (superherointj) [inline pkgs/applications/networking/browsers/mullvad-browser/default.nix:64]: Use `lib.optionalString`

File: pkgs/applications/networking/browsers/mullvad-browser/default.nix [hunk 8]
```diff
@@ -235,7 +261,19 @@ stdenv.mkDerivation rec {
     runHook postBuild
   '';
 
-  installPhase = ''
+  installPhase = if stdenv.isDarwin then ''
```
Comment (superherointj) [inline pkgs/applications/networking/browsers/mullvad-browser/default.nix:264]: `lib.optionalString` would accomplish same you are doing. But I'm finding this pattern substitution weird here. I'm afraid this might disable install hooks and re-write `installPhase` to empty. But needs to be confirmed.

**Comment by heywoodlh** [comment]
@superherointj sorry I missed the notifications for your comments! Thanks for the ping. Will get around to addressing them sometime tonight or tomorrow.

**Comment by Eveeifyeve** [comment]
Hey good news about this pr they recently added support for darwin in firefox in this pr: https://github.com/NixOS/nixpkgs/pull/350384
